core/commands/dht.go

```diff
@@ -39,13 +45,191 @@ var DhtCmd = &cmds.Command{
 )
 
+type dhtPeerInfo struct {
+	ID            string
+	Connected     bool
+	AgentVersion  string
+	LastUsefulAt  string
+	LastQueriedAt string
+}
+
+type dhtStat struct {
+	Name    string
+	Buckets []dhtBucket
+}
+
+type dhtBucket struct {
+	LastRefresh string
+	Peers       []dhtPeerInfo
+}
+
+var statDhtCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Returns stats about the DHTs",
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("dht", false, true, "The DHT whose table should be listed (wan or lan)."),
+	},
+	Options: []cmds.Option{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		nd, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		if !nd.IsOnline {
+			return ErrNotOnline
+		}
+
+		if nd.DHT == nil {
+			return ErrNotDHT
+		}
+
+		id := kbucket.ConvertPeerID(nd.Identity)
+
+		dhts := make(map[string]*dht.IpfsDHT)
+		if len(req.Arguments) == 0 {
+			dhts["wan"] = nd.DHT.WAN
+			dhts["lan"] = nd.DHT.LAN
+		} else {
+			for _, arg := range req.Arguments {
+				switch arg {
+				case "wan":
+				case "lan":
+				default:
+					return cmds.Errorf(cmds.ErrClient, "unknown dht type: %s", arg)
+				}
+			}
+		}
+
+		for name, dht := range dhts {
+			rt := dht.RoutingTable()
+			lastRefresh := rt.GetTrackedCplsForRefresh()
+			infos := rt.GetPeerInfos()
+			buckets := make([]dhtBucket, 0, len(lastRefresh))
+			for _, pi := range infos {
+				cpl := kbucket.CommonPrefixLen(id, kbucket.ConvertPeerID(pi.Id))
+				if len(buckets) <= cpl {
+					buckets = append(buckets, make([]dhtBucket, 1+cpl-len(buckets))...)
+				}
+
+				info := dhtPeerInfo{ID: pi.Id.String()}
+
+				if ver, err := nd.Peerstore.Get(pi.Id, "AgentVersion"); err == nil {
+					info.AgentVersion, _ = ver.(string)
+				} else if err == pstore.ErrNotFound {
+					// ignore
+				} else {
+					// this is a bug, usually.
+					log.Errorw(
+						"failed to get agent version from peerstore",
+						"error", err,
+					)
+				}
+				if !pi.LastUsefulAt.IsZero() {
+					info.LastUsefulAt = pi.LastUsefulAt.Format(time.RFC3339)
+				}
+
+				if !pi.LastSuccessfulOutboundQueryAt.IsZero() {
+					info.LastQueriedAt = pi.LastSuccessfulOutboundQueryAt.Format(time.RFC3339)
+				}
+
+				info.Connected = nd.PeerHost.Network().Connectedness(pi.Id) == network.Connected
+
+				buckets[cpl].Peers = append(buckets[cpl].Peers, info)
+			}
+			for i := 0; i < len(buckets) && i < len(lastRefresh); i++ {
+				refreshTime := lastRefresh[i]
+				if !refreshTime.IsZero() {
+					buckets[i].LastRefresh = refreshTime.Format(time.RFC3339)
+				}
+			}
+			res.Emit(dhtStat{
+				Name:    name,
+				Buckets: buckets,
+			})
+		}
+
+		return nil
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out dhtStat) error {
+			tw := tabwriter.NewWriter(w, 4, 4, 2, ' ', 0)
+			defer tw.Flush()
+
+			// Regex to remove fraction from duration.
+			re := regexp.MustCompile(`\.\d+`)
```

Comment:
declare as variable outside of function?

Reply:
Done.